File: tools/tf-types-ts/src/core/bridges.ts
/**
 * Common bridge framework. Concrete bridges (SPIFFE, WebAuthn, MCP) register
 * themselves with a BridgeRegistry so higher-level code can look up a bridge
 * by kind at runtime.
 */

export type BridgeKind =
  | "spiffe"
  | "webauthn"
  | "mcp"
  | "oauth"
  | "gnap"
  | "tls"
  | "did"
  | "matrix"
  | "webhook"
  | "grpc"
  | "service-mesh"
  | "a2a";

export interface BridgeError {
  code: "unsupported" | "invalid-input" | "rejected" | "internal";
  message: string;
}

export class BridgeFailure extends Error {
  readonly code: BridgeError["code"];
  constructor(err: BridgeError) {
    super(`${err.code}: ${err.message}`);
    this.code = err.code;
  }
}

export interface Bridge {
  readonly bridgeId: string;
  readonly kind: BridgeKind;
  readonly trustDomain: string;
}

export class BridgeRegistry {
  private bridges: Bridge[] = [];

  register(bridge: Bridge): void {
    this.bridges.push(bridge);
  }

  get<T extends Bridge>(kind: BridgeKind): T | undefined {
    return this.bridges.find((b) => b.kind === kind) as T | undefined;
  }

  list(): Bridge[] {
    return [...this.bridges];
  }
}

File: crates/tf-types/src/bridges.rs
//! Common compatibility-bridge framework. Concrete bridges live in
//! sibling modules (e.g. `bridge_spiffe`) and register themselves with a
//! `BridgeRegistry` so higher-level code can look up a bridge by kind at
//! runtime.

use std::sync::Arc;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum BridgeKind {
    Spiffe,
    Webauthn,
    Mcp,
    Oauth,
    Tls,
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum BridgeError {
    #[error("bridge unsupported: {0}")]
    Unsupported(String),
    #[error("bridge invalid input: {0}")]
    InvalidInput(String),
    #[error("bridge rejected input: {0}")]
    Rejected(String),
    #[error("bridge internal error: {0}")]
    Internal(String),
}

pub trait Bridge: Send + Sync {
    fn bridge_id(&self) -> &str;
    fn kind(&self) -> BridgeKind;
    fn trust_domain(&self) -> &str;
}

#[derive(Default)]
pub struct BridgeRegistry {
    bridges: Vec<Arc<dyn Bridge>>,
}

impl BridgeRegistry {
    pub fn new() -> Self {
        BridgeRegistry { bridges: Vec::new() }
    }

    pub fn register(&mut self, bridge: Arc<dyn Bridge>) {
        self.bridges.push(bridge);
    }

    pub fn get(&self, kind: BridgeKind) -> Option<Arc<dyn Bridge>> {
        self.bridges.iter().find(|b| b.kind() == kind).cloned()
    }

    pub fn list(&self) -> &[Arc<dyn Bridge>] {
        &self.bridges
    }
}
